ProcessList.c

```diff
@@ -194,308 +191,141 @@ void ProcessList_remove(ProcessList* this, const Process* p) {
-
-      int lft = (*node_counter)++;
 
       if (!show) {
          process->show = false;
       }
 
-      int s = Vector_size(this->processes2);
-      if (direction == 1) {
-         Vector_add(this->processes2, process);
-      } else {
-         Vector_insert(this->processes2, 0, process);
-      }
+      int s = Vector_size(this->displayList);
+      Vector_add(this->displayList, process);
 
-      assert(Vector_size(this->processes2) == s + 1); (void)s;
+      assert(Vector_size(this->displayList) == s + 1); (void)s;
 
       int nextIndent = indent | (1 << level);
-      ProcessList_buildTreeBranch(this, process->pid, level + 1, (i < size - 1) ? nextIndent : indent, direction, show ? process->showChildren : false, node_counter, node_index);
-      if (i == size - 1) {
+      ProcessList_buildTreeBranch(this, process->pid, level + 1, (i < lastShown) ? nextIndent : indent, process->show && process->showChildren);
+      if (i == lastShown) {
          process->indent = -nextIndent;
       } else {
          process->indent = nextIndent;
       }
 
-      int rht = (*node_counter)++;
-
-      process->tree_left = lft;
-      process->tree_right = rht;
       process->tree_depth = level + 1;
-      process->tree_index = index;
-      Hashtable_put(this->displayTreeSet, index, process);
    }
-   Vector_delete(children);
 }
 
-static int ProcessList_treeProcessCompare(const void* v1, const void* v2) {
+static int compareProcessByKnownParentThenNatural(const void* v1, const void* v2) {
    const Process* p1 = (const Process*)v1;
    const Process* p2 = (const Process*)v2;
 
-   return SPACESHIP_NUMBER(p1->tree_left, p2->tree_left);
-}
-
-static int ProcessList_treeProcessCompareByPID(const void* v1, const void* v2) {
-   const Process* p1 = (const Process*)v1;
-   const Process* p2 = (const Process*)v2;
+   int result = SPACESHIP_NUMBER(
+      p1->isRoot ? 0 : Process_getParentPid(p1),
+      p2->isRoot ? 0 : Process_getParentPid(p2)
+   );
+   if (result != 0)
+      return result;
 
-   return SPACESHIP_NUMBER(p1->pid, p2->pid);
+   return Process_compare(v1, v2);
 }
 
 // Builds a sorted tree from scratch, without relying on previously gathered information
 static void ProcessList_buildTree(ProcessList* this) {
-   int node_counter = 1;
-   int node_index = 0;
-   int direction = ScreenSettings_getActiveDirection(this->settings->ss);
+   Vector_prune(this->displayList);
 
-   // Sort by PID
-   Vector_quickSortCustomCompare(this->processes, ProcessList_treeProcessCompareByPID);
+   // Mark root processes
    int vsize = Vector_size(this->processes);
+   for (int i = 0; i < vsize; i++) {
+      Process* process = (Process*)Vector_get(this->processes, i);
+      pid_t ppid = Process_getParentPid(process);
+      process->isRoot = false;
+
+      // If PID corresponds with PPID (e.g. "kernel_task" (PID:0, PPID:0)
+      // on Mac OS X 10.11.6) regard this process as root.
+      if (process->pid == ppid)
+         process->isRoot = true;
```

Comment:
add `continue;` to skip `ProcessList_findProcess()`

Reply:
Done.